cmd/gpu-kubelet-plugin/driver.go

```diff
@@ -177,6 +195,112 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
+			if !ok {
+				klog.V(6).Info("Health monitor channel closed")
+				return
+			}
+
+			uuid := device.GetUUID()
+			klog.Warningf("Received unhealthy notification for device: %s", uuid)
+
+			// Mark device as unhealthy.
+			d.state.MarkDeviceUnhealthy(device)
+
+			// Update allocated device health status in a given claim
+			result, claimUID, err := d.findDeviceResultAndClaimUID(uuid)
+			if err != nil {
+				klog.Errorf("Device %s is unhealthy, but no associated claim was found: %v", uuid, err)
+				continue
+			}
+
+			claim, err := d.findClaimByUID(ctx, types.UID(claimUID))
+			if err != nil {
+				klog.Errorf("Failed to find ResourceClaim object for UID %s: %v", claimUID, err)
+				continue
+			}
+
+			ds := d.state.buildDeviceStatusApply(result, device.Health)
+			if err := d.state.applyClaimDeviceStatuses(ctx, claim.Namespace, claim.Name, ds); err != nil {
+				klog.Errorf("Failed to update status for claim %s/%s: %v", claim.Namespace, claim.Name, err)
+			} else {
+				klog.V(6).Infof("applied unhealthy device status to claim %s/%s", claim.Namespace, claim.Name)
+			}
+		}
+	}
+}
+
+func (d *driver) findDeviceResultAndClaimUID(unhealthyDeviceUUID string) (*resourceapi.DeviceRequestAllocationResult, string, error) {
+	d.state.Lock()
+	defer d.state.Unlock()
+
+	checkpoint, err := d.state.getCheckpoint()
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to get checkpoint: %v", err)
+	}
+
+	for claimUID, preparedClaim := range checkpoint.V2.PreparedClaims {
+		var matchingDeviceName string
+
+	SearchDeviceGroups:
+		for _, group := range preparedClaim.PreparedDevices {
+			for _, device := range group.Devices {
+				var currentUUID string
+				var currentDeviceName string
+
+				if device.Gpu != nil {
+					currentUUID = device.Gpu.Info.UUID
+					currentDeviceName = device.Gpu.Device.DeviceName
+				} else if device.Mig != nil {
+					currentUUID = device.Mig.Info.UUID
+					currentDeviceName = device.Mig.Device.DeviceName
+				}
+
+				if currentUUID == unhealthyDeviceUUID {
+					klog.V(6).Infof("found matching device: %v for claim: %s", currentDeviceName, claimUID)
+					matchingDeviceName = currentDeviceName
+					break SearchDeviceGroups
+				}
+			}
+		}
+
+		if matchingDeviceName != "" {
+			if preparedClaim.Status.Allocation != nil {
+				for i := range preparedClaim.Status.Allocation.Devices.Results {
+					result := &preparedClaim.Status.Allocation.Devices.Results[i]
+					if result.Device == matchingDeviceName {
+						klog.V(6).Infof("Found device result: %v for claim: %s", result.Device, claimUID)
+						return result, claimUID, nil
+					}
+				}
+			}
+		}
+	}
+	return nil, "", fmt.Errorf("unable to find device result and claim uid for %s", unhealthyDeviceUUID)
+}
+
+func (d *driver) findClaimByUID(ctx context.Context, claimUID types.UID) (*v1beta1.ResourceClaim, error) {
+	claimList, err := d.state.config.clientsets.Core.ResourceV1beta1().ResourceClaims("").List(ctx, metav1.ListOptions{})
```

Comment:
I'm not sure how I feel about listing all Resource claims and then searching through them for the matching UID. It feels like we should rather be storing the claim name / namespace in the checkpoint so we can pull it directly and then assert that it has the correct UID.

Reply:
Yes yes, this is not efficient, i dint want to make changes to the existing code as we are not sure if we want to take this action or not. This is more of a sample action.